source/common/upstream/health_discovery_service.cc

```diff
@@ -165,69 +164,129 @@ void HdsDelegate::onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&& metadata
   ASSERT(message);
+  std::vector<HdsClusterPtr> hds_clusters;
+  // Maps to replace the current member variable versions.
+  absl::flat_hash_map<std::string, HdsClusterPtr> new_hds_clusters_name_map;
 
   for (const auto& cluster_health_check : message->cluster_health_checks()) {
-    // Create HdsCluster config
-    static const envoy::config::core::v3::BindConfig bind_config;
-    envoy::config::cluster::v3::Cluster cluster_config;
-
-    cluster_config.set_name(cluster_health_check.cluster_name());
-    cluster_config.mutable_connect_timeout()->set_seconds(ClusterTimeoutSeconds);
-    cluster_config.mutable_per_connection_buffer_limit_bytes()->set_value(
-        ClusterConnectionBufferLimitBytes);
-
-    // Add endpoints to cluster
-    for (const auto& locality_endpoints : cluster_health_check.locality_endpoints()) {
-      // add endpoint group by locality to config
-      auto* endpoints = cluster_config.mutable_load_assignment()->add_endpoints();
-      // if this group contains locality information, save it.
-      if (locality_endpoints.has_locality()) {
-        endpoints->mutable_locality()->MergeFrom(locality_endpoints.locality());
-      }
-
-      // add all endpoints for this locality group to the config
-      for (const auto& endpoint : locality_endpoints.endpoints()) {
-        endpoints->add_lb_endpoints()->mutable_endpoint()->mutable_address()->MergeFrom(
-            endpoint.address());
-      }
+    HdsClusterPtr cluster_ptr;
+
+    // Create a new configuration for a cluster based on our different or new config.
+    auto cluster_config = createClusterConfig(cluster_health_check);
+
+    // If this particular cluster configuration happens to have a name, then it is possible
+    // this particular cluster exists in the name map. We check and if we found a match,
+    // attempt to update this cluster. If no match was found, either the cluster name is empty
+    // or we have not seen a cluster by this name before. In either case, create a new cluster.
+    auto cluster_map_pair = hds_clusters_name_map_.find(cluster_health_check.cluster_name());
+    if (cluster_map_pair != hds_clusters_name_map_.end()) {
+      // We have a previous cluster with this name, update.
+      cluster_ptr = cluster_map_pair->second;
+      updateHdsCluster(cluster_ptr, cluster_config);
+    } else {
+      // There is no cluster with this name previously or its an empty string, so just create a
+      // new cluster.
+      cluster_ptr = createHdsCluster(cluster_config);
     }
 
-    // TODO(lilika): Add support for optional per-endpoint health checks
-
-    // Add healthchecks to cluster
-    for (auto& health_check : cluster_health_check.health_checks()) {
-      cluster_config.add_health_checks()->MergeFrom(health_check);
+    // If this had a non-empty name, add this cluster to the name map so it can be updated in the
+    // future.
+    if (!cluster_health_check.cluster_name().empty()) {
```

Comment:
Empty cluster_name is unexpected, but I think it's OK to leave it as-is.
If there is really an empty cluster_name, skipping here only will cause HdsCluster being recreated every time.

Reply:
Exactly. Unfortunately the cluster_name() field has no validator in its proto saying that this field must not be empty, and AFAIK adding this is considered a breaking change and cannot be enforced until v4. This check here is intentional since a cluster without a name is unexpected and then it will just default to recreation since it will never be in a map.

Reply:
Discussed this more with @xdzhai, it is true that a empty string will work in a map, but I put this check here because the empty string is the default value, and cluster_name() was previously only used for reporting purposes and not required. Some user could have known this and decided to never set a cluster name, and they may have for example 100 clusters. If this check was not here after this change, then such a user would go from having 100 clusters to only 1.

I'm thinking we can't enforce users to have a cluster_name() since it was optional previously, so such users will get the old behavior, specifically where we recreated HdsClusters every time.

In the case that there are two clusters with the same name and are non-empty, those were manually set so I would think that to be a user error, and not something we should worry about. I think I should just add some ENVOY_LOG(warn, ...) messages for both of these cases and I think that should be enough.

Reply:
@htuch What are your thoughts on this?

Reply:
Yeah, I believe we are not able to change the assumptions around empty cluster names, as much sense as it does make to restrict to non-empty. I think we'll need the special casing. A warning log is fine.